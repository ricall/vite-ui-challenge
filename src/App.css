header {
  font-size: 24px;
  font-style: normal;
  font-weight: 700;
  line-height: normal;
}

.App {
  display: flex;
  text-align: center;
  flex-direction: column;
  align-items: center;
  gap: 20px;
}

.AppContainer {
  max-width: 1280px;

  display: flex;
  flex-direction: column;
  gap: 10px;
  align-items: center;
}
