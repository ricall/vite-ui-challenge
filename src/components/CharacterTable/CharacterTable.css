.CharacterTable {
  color: #000;
  font-size: 20px;
  font-style: normal;
  font-weight: 700;
  line-height: normal;
  border-collapse: collapse;
  margin: 40px;

  max-width: 1024px;
}

.CharacterTable th,
td {
  padding: 10px;
}

.CharacterTable tbody {
  border-radius: 10px;
  border: 0px solid #000;
  background: #fff;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.25);
}

.HeaderLeft {
  text-align: left;
}
