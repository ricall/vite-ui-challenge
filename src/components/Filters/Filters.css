.Filters {
  display: flex;
  padding: 0 20px;
  flex-wrap: wrap;
  flex-direction: row;
  justify-items: center;
  align-items: center;
  gap: 6px;
  margin: 0 40px;
}

.Clear {
  color: #999;
  margin-left: 10px;
  text-decoration: underline;
}
