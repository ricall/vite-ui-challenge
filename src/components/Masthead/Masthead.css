.Masthead {
  display: flex;
  position: relative;
  text-align: center;
  width: 100%;
  height: 122px;
}

.Masthead--header {
  position: absolute;
  width: 100%;
  height: 76px;
  background-color: #000;
}

.Masthead-icon {
  position: absolute;
  width: 160px;
  height: 90px;
  top: 32px;
  left: 50%;
  transform: translate(-50%, 0);
  background: url(../../img/Mortal-Kombat-Logo.png) no-repeat;
  background-size: cover;
}
