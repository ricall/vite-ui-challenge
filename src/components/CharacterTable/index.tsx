import { CharacterRow } from './CharacterRow';
import './CharacterTable.css';
import { useCharacterStore } from '../../store/characters.ts';

export function CharacterTable() {
  const { getFilteredCharacters, isOnTeam, toggleTeamMember } = useCharacterStore();

  const characters = getFilteredCharacters();

  const handleSelect = (id: number) => () => toggleTeamMember(id);

  return (
    <table className="CharacterTable">
      <thead>
        <tr>
          <th colSpan={3} className="HeaderLeft">
            Character
          </th>
          <th className="HeaderLeft">Tags</th>
          <th>Power</th>
          <th>Mobility</th>
          <th>Technique</th>
          <th>Survivability</th>
          <th>Energy</th>
        </tr>
      </thead>
      <tbody>
        {characters.map(({ id, ...character }) => (
          <CharacterRow id={id} key={id} selected={isOnTeam(id)} onToggleTeamMember={handleSelect(id)} {...character} />
        ))}
      </tbody>
    </table>
  );
}
