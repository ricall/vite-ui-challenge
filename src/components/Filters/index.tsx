import { tagNames } from '../../model';
import './Filters.css';
import { TagBadge } from '../TagBadge';
import { useCharacterStore } from '../../store/characters.ts';

export function Filters() {
  const { isTagSelected, toggleTag, setTags } = useCharacterStore();

  const handleToggleTag = (tag: string) => () => toggleTag(tag);
  const handleClearAll = () => setTags([]);

  return (
    <div className="Filters">
      {tagNames.map((tag) => (
        <TagBadge key={tag} label={tag} isChecked={isTagSelected(tag)} onClick={handleToggleTag(tag)} />
      ))}
      <div className="Clear" onClick={handleClearAll}>
        Clear all
      </div>
    </div>
  );
}
