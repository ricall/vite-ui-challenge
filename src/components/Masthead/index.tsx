import './Masthead.css';

export function Masthead() {
  return (
    <div className="Masthead">
      <div className="Masthead--header" />
      <div className="Masthead-icon" />
    </div>
  );
}
