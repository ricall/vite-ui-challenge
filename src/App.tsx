import './App.css';
import { CharacterTable } from './components/CharacterTable';
import { Filters } from './components/Filters';
import { Masthead } from './components/Masthead';
import { Search } from './components/Search';
import { Team } from './components/Team';
import { Totals } from './components/Totals';

function App() {
  return (
    <div className="App">
      <Masthead />
      <div className="AppContainer">
        <header>Your champions!</header>
        <Team />
        <Totals />
        <Search />
        <Filters />
        <CharacterTable />
      </div>
    </div>
  );
}

export default App;
