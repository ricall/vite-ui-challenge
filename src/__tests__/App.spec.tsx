import { render } from '@testing-library/react';

import App from '../App';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { Masthead } from '../components/Masthead';
import { Team } from '../components/Team';
import { Totals } from '../components/Totals';
import { Search } from '../components/Search';
import { Filters } from '../components/Filters';
import { CharacterTable } from '../components/CharacterTable';

vi.mock('../components/Masthead');
vi.mock('../components/Team');
vi.mock('../components/Totals');
vi.mock('../components/Search');
vi.mock('../components/Filters');
vi.mock('../components/CharacterTable');

describe('App', () => {
  const mockComponent = (name: string) => <div data-testid={name} />;
  beforeEach(() => {
    vi.resetAllMocks();

    vi.mocked(Masthead).mockReturnValue(mockComponent('masthead'));
    vi.mocked(Team).mockReturnValue(mockComponent('team'));
    vi.mocked(Totals).mockReturnValue(mockComponent('totals'));
    vi.mocked(Search).mockReturnValue(mockComponent('search'));
    vi.mocked(Filters).mockReturnValue(mockComponent('filters'));
    vi.mocked(CharacterTable).mockReturnValue(mockComponent('table'));
  });

  it('should render the app as expected', () => {
    const { getByText, getByTestId, asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
    expect(getByText('Your champions!')).toBeDefined();
    ['masthead', 'team', 'totals', 'search', 'filters', 'table'].forEach((name) =>
      expect(getByTestId(name)).toBeDefined(),
    );
  });
});
