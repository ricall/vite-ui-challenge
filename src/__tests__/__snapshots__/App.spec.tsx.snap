// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render the app as expected 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      data-testid="masthead"
    />
    <div
      class="AppContainer"
    >
      <header>
        Your champions!
      </header>
      <div
        data-testid="team"
      />
      <div
        data-testid="totals"
      />
      <div
        data-testid="search"
      />
      <div
        data-testid="filters"
      />
      <div
        data-testid="table"
      />
    </div>
  </div>
</DocumentFragment>
`;
